File: Data Structures & Algorithms/longest-palindromic-substring/submission-0.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        if (s.length() < 1) return "";
        int start = 0, end = 0;

        for (int i = 0; i < s.length(); i++) {
            // Case 1: Window centered on a single character (aba)
            int len1 = expandFromCenter(s, i, i);
            // Case 2: Window centered between two characters (abba)
            int len2 = expandFromCenter(s, i, i + 1);
            
            int len = max(len1, len2);
            if (len > end - start) {
                // Update our global "best" window bounds
                start = i - (len - 1) / 2;
                end = i + len / 2;
            }
        }
        return s.substr(start, end - start + 1);
    }

private:
    int expandFromCenter(string s, int left, int right) {
        // Expand the window as long as it's a palindrome
        while (left >= 0 && right < s.length() && s[left] == s[right]) {
            left--;
            right++;
        }
        // Return the length of the palindrome found
        return right - left - 1;
    }
};

File: Data Structures & Algorithms/trapping-rain-water/submission-0.py
class Solution:
    def trap(self, height: List[int]) -> int:
        
        left = 0
        right = len(height) - 1
        max_left = height[left]
        max_right = height[right]
        total_vol = 0

        while (left < right):
            if(max_right < max_left):
                right -= 1
                max_right = max(max_right, height[right])
                total_vol += max_right - height[right]
            else:
                left += 1
                max_left = max(max_left, height[left])
                total_vol += max_left - height[left]
        
        return total_vol



        